### PR TITLE
Extend thread state logging and change default stealing parameters

### PR DESCRIPTION
**Pull request by msimberg**
This PR is the result of some debuggin in DLA-Future of a deadlock caused by a combination of our scheduler and a blocking MPI call being made in an HPX thread. We know that the latter is very naughty, but at the same time it can lead to performance benefits in this particular case (lots of small tasks if one does not do the blocking call). We're looking into alternatives, but one of the changes here also helps for the particular situation. I'm happy to drop it though as it's only a question of default configuration values. That said, the changes here are:

- Add more extensive logging to the scheduling loop. Previously the state of an HPX thread was logged only once the context was changed back to the scheduler. This PR adds logging both before and after threads are scheduled so that one can see when they start and stop. This helps with recognizing if a task has been started but never finishes. I've also added more information to the messages (thread pool, scheduler). I also log when attmpting to steal threads, but the number of threads is lower than the threshold.
- Change the `idle_loop_count` to count up again like the other counters. The counter was decremented in most places, but incremented in another (https://github.com/STEllAR-GROUP/hpx/blob/bb51c2c8cbd0dc68026f3bd92b0b746b7a273d86/libs/core/schedulers/include/hpx/schedulers/queue_helpers.hpp#L53). This alone did not cause a deadlock, but it was part of it. The `idle_loop_count` never reached a value where `enable_stealing_staged` would be true. @hkaiser I know you changed this to count down in https://github.com/STEllAR-GROUP/hpx/pull/3768. However, I don't see which cases become more cumbersome when incrementing the counter (https://github.com/STEllAR-GROUP/hpx/pull/3768#discussion_r272145021). Do you remember what those were? Does `idle_loop_count` need to be reset to `max_idle_loop_count` somewhere outside the scheduling loop?
- Finally, change default value of `min_tasks_to_steal_staged` to 0. This is not a must, mostly convenience.

### CODE REVIEW

**Comment by hkaiser** [comment]
> This PR is the result of some debuggin in DLA-Future of a deadlock caused by a combination of our scheduler and a blocking MPI call being made in an HPX thread. We know that the latter is very naughty, but at the same time it can lead to performance benefits in this particular case (lots of small tasks if one does not do the blocking call). We're looking into alternatives, but one of the changes here also helps for the particular situation. I'm happy to drop it though as it's only a question of default configuration values. That said, the changes here are:
> 
> * Add more extensive logging to the scheduling loop. Previously the state of an HPX thread was logged only once the context was changed back to the scheduler. This PR adds logging both before and after threads are scheduled so that one can see when they start and stop. This helps with recognizing if a task has been started but never finishes. I've also added more information to the messages (thread pool, scheduler). I also log when attempting to steal threads, but the number of threads is lower than the threshold.

Very nice. Thanks!

> * Change the `idle_loop_count` to count up again like the other counters. The counter was decremented in most places, but incremented in another (https://github.com/STEllAR-GROUP/hpx/blob/bb51c2c8cbd0dc68026f3bd92b0b746b7a273d86/libs/core/schedulers/include/hpx/schedulers/queue_helpers.hpp#L53). 

This is a bug. I shouldn't ever have been incremented.

> This alone did not cause a deadlock, but it was part of it. The `idle_loop_count` never reached a value where `enable_stealing_staged` would be true. @hkaiser I know you changed this to count down in #3768. However, I don't see which cases become more cumbersome when incrementing the counter ([#3768 (comment)](https://github.com/STEllAR-GROUP/hpx/pull/3768#discussion_r272145021)). Do you remember what those were? Does `idle_loop_count` need to be reset to `max_idle_loop_count` somewhere outside the scheduling loop?

IIRC, I changed it to count down as it needed to be tested some place where the `max_idle_loop_count` wasn't readily available. But this might have been changed since then (sorry I don't remember the details). I'm fine with changing it back to count up if there is no problem doing so.

> * Finally, change default value of `min_tasks_to_steal_staged` to 0. This is not a must, mostly convenience.

I admit that at some point I did run into a deadlock issue caused by this value not being zero as well and forgot to change the default value as this would have prevented my deadlock (I did create an issue, though: #4558). Thanks for this.

File: libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp [hunk 1]
```diff
@@ -1004,11 +1004,24 @@ namespace hpx { namespace threads { namespace policies {
...
-                            std::memory_order_relaxed))
+                std::int64_t new_tasks_count =
+                    addfrom->new_tasks_count_.data_.load(
+                        std::memory_order_relaxed);
+                bool enough_threads =
+                    new_tasks_count >= parameters_.min_tasks_to_steal_staged_;
+
+                if (running && !enough_threads)
                 {
+                    if (!enough_threads && new_tasks_count != 0)
```
Comment (hkaiser) [inline libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp:1015]:
```suggestion
                    if (new_tasks_count != 0)
```

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by albestro** [comment]
Just tested again, and it works! and the new debug log is super useful.

thanks @msimberg!

**Comment by msimberg** [comment]
> This is a bug. I shouldn't ever have been incremented.
> IIRC, I changed it to count down as it needed to be tested some place where the `max_idle_loop_count` wasn't readily available. But this might have been changed since then (sorry I don't remember the details). I'm fine with changing it back to count up if there is no problem doing so.

No worries. I am slightly worried that I'll introduce the same bug in the opposite direction with this change, so I'll have another look through the code to see if I find a case like this. My reasoning for changing the direction is that it's consistent with the counts elsewhere and thus it'll hopefully avoid a similar bug in the future, but that doesn't mean I can't mess it up while doing this change...

> I admit that at some point I did run into a deadlock issue caused by this value not being zero as well and forgot to change the default value as this would have prevented my deadlock (I did create an issue, though: #4558). Thanks for this.

Ok, sounds good. Users will still have the choice of increasing it if they know it's not worth it because of too small tasks, but this sounds like a better default in that case.

**Comment by msimberg** [comment]
> No worries. I am slightly worried that I'll introduce the same bug in the opposite direction with this change, so I'll have another look through the code to see if I find a case like this. My reasoning for changing the direction is that it's consistent with the counts elsewhere and thus it'll hopefully avoid a similar bug in the future, but that doesn't mean I can't mess it up while doing this change...

Well, to the best of my knowledge there are no cases like this (anymore at least). I'll go ahead and merge this now.

**Pull request merged**